src/app/codegen-data-model/CodegenDataModel_Read.cpp

```diff
@@ -0,0 +1,350 @@
+    case ZCL_LONG_OCTET_STRING_ATTRIBUTE_TYPE:
+        return EncodeStringLike<ByteSpan, LongPascalString>(data, isNullable, encoder);
+    default:
+        ChipLogError(DataManagement, "Attribute type 0x%x not handled", static_cast<int>(metadata->attributeType));
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedRead);
+    }
+}
+
+} // namespace
+
+/// separated-out ReadAttribute implementation (given existing complexity)
+///
+/// Generally will:
+///    - validate ACL (only for non-internal requests)
+///    - Try to read attribute via the AttributeAccessInterface
+///    - Try to read the value from ember RAM storage
+CHIP_ERROR CodegenDataModel::ReadAttribute(const InteractionModel::ReadAttributeRequest & request, AttributeValueEncoder & encoder)
+{
+    ChipLogDetail(DataManagement,
+                  "Reading attribute: Cluster=" ChipLogFormatMEI " Endpoint=%x AttributeId=" ChipLogFormatMEI " (expanded=%d)",
+                  ChipLogValueMEI(request.path.mClusterId), request.path.mEndpointId, ChipLogValueMEI(request.path.mAttributeId),
+                  request.path.mExpanded);
+
+    // ACL check for non-internal requests
+    if (!request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+        ReturnErrorCodeIf(!request.subjectDescriptor.has_value(), CHIP_ERROR_INVALID_ARGUMENT);
+
+        Access::RequestPath requestPath{ .cluster = request.path.mClusterId, .endpoint = request.path.mEndpointId };
+        ReturnErrorOnFailure(Access::GetAccessControl().Check(*request.subjectDescriptor, requestPath,
+                                                              RequiredPrivilege::ForReadAttribute(request.path)));
+    }
+
+    auto metadata = FindAttributeMetadata(request.path);
+
+    // Explicit failure in finding a suitable metadata
+    if (const CHIP_ERROR * err = std::get_if<CHIP_ERROR>(&metadata))
+    {
+        VerifyOrDie(*err != CHIP_NO_ERROR);
+        return *err;
+    }
+
+    // Read via AAI
+    std::optional<CHIP_ERROR> aai_result;
+    if (const EmberAfCluster ** cluster = std::get_if<const EmberAfCluster *>(&metadata))
+    {
+        Compatibility::GlobalAttributeReader aai(*cluster);
+        aai_result = TryReadViaAccessInterface(request.path, &aai, encoder);
+    }
+    else
+    {
+        aai_result = TryReadViaAccessInterface(
+            request.path, GetAttributeAccessOverride(request.path.mEndpointId, request.path.mClusterId), encoder);
+    }
+    ReturnErrorCodeIf(aai_result.has_value(), *aai_result);
+
+    if (!std::holds_alternative<const EmberAfAttributeMetadata *>(metadata))
+    {
+        // if we only got a cluster, this was for a global attribute. We cannot read ember attributes
+        // at this point, so give up (although GlobalAttributeReader should have returned something here).
+        // Return a permanent failure...
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
```

Comment:
Is there a reason this is not a VerifyOrDie? That said, you already hit the relevant VerifyOrDieWithMsg in GlobalAttributeReader::Read, no?  This line is unreachable...

Reply:
This method does not technically have visibility in GlobalAttributeReader::Read ... it needs to work as a stand-alone. This seemed appropriate but I can do VerifyOrDie as well.